contributors/design-proposals/network/service_port_ranges.md

```diff
@@ -0,0 +1,128 @@
+
+## Goals
+
+A server side solution to support single or multiple contiguous port range.
+
+## Non-goal
+
+Target port remapping in ranges.
+
+Pod port ranges.
+
+## User Cases
+
+* Specifying port ranges in services declarations is mandatory for any VoIP-related application.
+* With the standard RTP port range which is 16384-32767. Clients connect on the standard RTP port range with a media server application.
+* Needs this change to make TeleStax (an RTC framework) work well in Kubernetes.
+
+## Proposal
+
+This proposal builds off earlier feature requrest to [Support port ranges in services](https://github.com/kubernetes/kubernetes/issues/23864). We aims to implement service port ranges at least for iptables and IPVS modes.
+
+### Service API changes
+
+Create a new filed `portRangeSize` in `ServicePort`.
+
+
+```go
+type ServiceSpec struct {
+  Ports []ServicePort
+}
+
+type ServicePort struct {
+  // portRangeSize is the contiguous ports number that are exposed by this service. Given port = 80 and portRangeSize = 100, it means ports 80-179, 100 ports in total, will be exposed. In this case, port means the starting port of a range.
+  // portRangeSize is default to be 1 for compatibility.
+  // If nodePort is specified, then nodePort is the starting port of range and nodePort+portRangeSize-1 is the ending port of range.
+  // A service can expose multiple port ranges and each port range should not be overlapped.
+  // Service port range does not support target port remapping, it implies that targetPort should not be specified or should be equal to service port.
+  // +optional
+  PortRangeSize int `json: "portRangeSize,omitempty"`
+}
+```
+
+### Service port ranges example
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: service-port-ranges
+  namespace: default
+spec:
+  ports:
+  - name: range1
+    port: 90
+    protocol: TCP
+    portRangeSize: 100
+    protocol: UDP
+  - name: range2
+    port: 200
+    portRange: 300
+    protocol: TCP
+```
+
+### Kube-proxy implementation limitations
+
+NodePort and ExternalIPs Services don't support port ranges since kube-proxy will hold local ports open for these services(nodeport and "external" IP happens to be an IP that is local to this machine). It's challenging to keep some huge port ranges open on host.
+
+Target port remapping is not supported since both iptables and ipvs does not do port translations in the mean time of doing port ranges.
+
+Userspace proxy mode in both linux and windows platform does not work with port ranges. Kube-proxy running in these modes will log error message if people create a service with `portRangeSize` > 1.
```

Comment:
As mentioned on https://github.com/kubernetes/kubernetes/issues/23864#issuecomment-382029019, there are probably workable solutions for the linux userspace proxy. I don't know about Windows. (There is also a Windows kernel proxy, which isn't mentioned here, and which I also don't know about.)

Reply:
@khenidak comment on the Windows side?

Reply:
@PatrickLang -- Can you help answering this?

Reply:
@dineshgovindasamy or @madhanrm - can you help review this? Can you forward a group of ports, or does it makes sense to unroll the ranges into discrete calls from kube-proxy?


cc @jmesser81 @daschott